frame/balances/src/lib.rs

```diff
@@ -1148,59 +1109,74 @@ impl<T: Trait<I>, I: Instance> Currency<T::AccountId> for Module<T, I> where
 	) -> Self::PositiveImbalance {
 		if value.is_zero() { return Self::PositiveImbalance::zero() }
 
-		let old_account = Self::account(who);
-		let mut account = old_account.clone();
-		let ed = T::ExistentialDeposit::get();
+		Self::try_mutate_account(who, |account| -> Result<Self::PositiveImbalance, Self::PositiveImbalance> {
+			// bail if not yet created and this operation wouldn't be enough to create it.
+			let ed = T::ExistentialDeposit::get();
+			ensure!(value >= ed || !account.total().is_zero(), Self::PositiveImbalance::zero());
+
+			// defensive only: overflow should never happen, however in case it does, then this
+			// operation is a no-op.
+			account.free = account.free.checked_add(&value).ok_or(Self::PositiveImbalance::zero())?;
+
+			Ok(PositiveImbalance::new(value))
+		}).unwrap_or_else(|x| x)
+	}
+
+	/// Withdraw some free balance from an account, respecting existence requirements.
+	///
+	/// Is a no-op if value to be withdrawn is zero.
+	fn withdraw(
+		who: &T::AccountId,
+		value: Self::Balance,
+		reasons: WithdrawReasons,
+		liveness: ExistenceRequirement,
+	) -> result::Result<Self::NegativeImbalance, DispatchError> {
+		if value.is_zero() { return Ok(NegativeImbalance::zero()); }
+
+		Self::try_mutate_account(who, |account|
+			-> Result<Self::NegativeImbalance, DispatchError>
+		{
+			let new_free_account = account.free.checked_sub(&value)
+				.ok_or(Error::<T, I>::InsufficientBalance)?;
 
-		// bail if not yet created and this operation wouldn't be enough to create it.
-		if value < ed && account.total().is_zero() { return Self::PositiveImbalance::zero() }
+			// bail if we need to keep the account alive and this would kill it.
+			let ed = T::ExistentialDeposit::get();
+			let would_kill = new_free_account < ed && account.free >= ed;
```

Comment:
We need to look at total balance for all the ED checks.

Reply:
Would be good to search for all usage of ED, and check it is using total balance for comparison

Reply:
done.